### PR TITLE
IPROTO-322 Schema compatibility checks

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/IPROTO-322

based on https://github.com/infinispan/protostream/pull/256

### CODE REVIEW

File: core/src/test/java/org/infinispan/protostream/ProtoLockTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            "org/infinispan/protostream/test/message.proto",
+            "sample_bank_account/bank.proto"
+      )) {
+         fileDescriptorSource.addProtoFile(f, asFile(f));
+      }
+      Map<String, FileDescriptor> parseResult = parseAndResolve(fileDescriptorSource, Configuration.builder().build());
+      ProtoLock protoLock1 = new ProtoLock(parseResult.values());
+      protoLock1.writeLockFile(baos);
+      String s = baos.toString(StandardCharsets.UTF_8);
+      System.out.println(s);
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/protostream/ProtoLockTest.java:39]: Remove?

File: core/src/main/java/org/infinispan/protostream/descriptors/ProtoLock.java [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+/**
+ * Reads/writes <a href="https://github.com/nilslice/protolock">protolock</a> files.
+ */
+public class ProtoLock {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/protostream/descriptors/ProtoLock.java:27]: There's a a few bits of duplicated code between the `writeMessage` and `writeEnum` being highlighted by Intellij, we should extract these to methods.

File: core/src/main/java/org/infinispan/protostream/descriptors/EnumDescriptor.java [hunk 3]
```diff
@@ -151,6 +153,53 @@ public String toString() {
...
+            if (!that.reservedNumbers.get(thisValue.getNumber())) {
+               errors.add(Log.LOG.removedFieldNotReserved(thisValue.getFullName(), thisValue.getNumber()));
+            }
+         } else {
+            if (thisValue.getNumber() != thatValue.getNumber()) {
+               errors.add(Log.LOG.modifiedFieldNumber(thisValue.getFullName(), thisValue.getNumber(), thatValue.getNumber()));
+            }
+         }
+      }
+      if (strict) {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/protostream/descriptors/EnumDescriptor.java:189]: This is the same as in the `Descriptor` class, maybe we should extract this to a method in `AnnotatedDescriptorImpl`?

File: core/src/test/java/org/infinispan/protostream/EvolutionTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import static org.infinispan.protostream.impl.parser.DescriptorsTest.asFile;
+import static org.infinispan.protostream.impl.parser.DescriptorsTest.parseAndResolve;
+
+import java.io.IOException;
+
+import org.infinispan.protostream.config.Configuration;
+import org.infinispan.protostream.descriptors.FileDescriptor;
+import org.junit.Test;
+
+public class EvolutionTest {
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/protostream/EvolutionTest.java:13]: `SchemaEvolutionTest`?

**Review by ryanemerson — COMMENTED** [review]
LGTM, just a few very minor comments.

File: core/src/test/java/org/infinispan/protostream/EvolutionTest.java (hunk 4 above)
Comment (tristantarrant) [reply core/src/test/java/org/infinispan/protostream/EvolutionTest.java]: Renamed

File: core/src/main/java/org/infinispan/protostream/descriptors/EnumDescriptor.java (hunk 3 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/protostream/descriptors/EnumDescriptor.java]: Refactored

File: core/src/main/java/org/infinispan/protostream/descriptors/ProtoLock.java (hunk 2 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/protostream/descriptors/ProtoLock.java:27]: Done

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**